### PR TITLE
feat: add skeleton for uniswapv2router tests

### LINKED ISSUE #1123: dev: add a test for UniswapV2Router02.sol 's addLiquidity entrypoint
@ClementWalter need help scoping

@danilowhk has succeeded to deploy UniswapV2Router02 on Kakarot Sepolia: ✅ 
@danilowhk has succeeded to add a pair on Kakarot Sepolia: ✅ 
@danilowhk has failed to add liquidity on Kakarot Sepolia: ❌

Proposal:
Add UniswapV2Router02.sol in the solidity contracts
Add the flow that Danilo described in a e2e test

Error message:
```
Am running out of steps when trying to add liquidity on router02:

transact to UniswapV2Router02.addLiquidity pending ... 
transact to UniswapV2Router02.addLiquidity errored: Error occurred: kakarot error: Provider errror StarknetError(ContractError(ContractErrorData { revert_error: "Error in the called contract (0x070c14f7fe5968975cff5cc4614c9d9a8cbb9051da19a28990772e3b55c23328):\nError at pc=0:5244:\nCould not reach the end of the program. RunResources has no remaining steps.\nCairo traceback (most recent call last):\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22756)\nUnknown location (pc=0:22668)\nUnknown location (pc=0:18719)\n" }))..

kakarot error: Provider errror StarknetError(ContractError(ContractErrorData { revert_error: "Error in the called contract (0x070c14f7fe5968975cff5cc4614c9d9a8cbb9051da19a28990772e3b55c23328):\nError at pc=0:5244:\nCould not reach the end of the program. RunResources has no remaining steps.\nCairo traceback (most recent call last):\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22758)\nUnknown location (pc=0:22756)\nUnknown location (pc=0:22668)\nUnknown location (pc=0:18719)\n" })).

Debug the transaction to get more information.
```

### PR DESCRIPTION
**Pull request by Eikix**
Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #1123 

## What is the new behavior?

- new tests covering uniswapv2 router

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1124)

### CODE REVIEW

File: tests/end_to_end/UniswapV2/conftest.py [hunk 1]
```diff
@@ -22,6 +22,14 @@ async def token_b(deploy_solidity_contract, owner):
         caller_eoa=owner.starknet_contract,
     )
 
+@pytest_asyncio.fixture(scope="module")
+async def token_WETH(deploy_solidity_contract, owner):
```
Comment (Eikix) [inline tests/end_to_end/UniswapV2/conftest.py:26]: I don't think this is WETH?

File: tests/end_to_end/UniswapV2/conftest.py [hunk 2]
```diff
@@ -34,3 +42,18 @@ async def factory(
...
+
+@pytest_asyncio.fixture(scope="module")
+async def router(
+    deploy_solidity_contract,
+    owner,
+    factory,
+    token_WETH,
+):
+    return await deploy_solidity_contract(
+        "UniswapV2",
```
Comment (Eikix) [inline tests/end_to_end/UniswapV2/conftest.py:54]: This won't work

File: .python-version
Comment (ClementWalter) [inline .python-version:1]: this is a pyenv local config file, remove

File: Makefile [hunk 3]
```diff
@@ -68,4 +68,4 @@ install-katana:
 	cargo install --git https://github.com/dojoengine/dojo --locked --tag "${KATANA_VERSION}" katana
 
 run-katana:
-	katana --chain-id test --validate-max-steps 4000000 --invoke-max-steps 6000000 --eth-gas-price 0 --strk-gas-price 0 --disable-fee
+	katana --chain-id test --validate-max-steps 160000000 --invoke-max-steps 160000000 --eth-gas-price 0 --strk-gas-price 0 --disable-fee
```
Comment (ClementWalter) [inline Makefile:71]: but we merged recently the opposite to be iso prod?

File: solidity_contracts/src/UniswapV2/UniswapV2ERC20.sol [hunk 4]
```diff
@@ -87,6 +87,11 @@ contract UniswapV2ERC20 is IUniswapV2ERC20 {
         return true;
     }
 
+    function mint(address to, uint value) external returns (bool) {
+        _mint(to, value);
+        return true;
```
Comment (ClementWalter) [inline solidity_contracts/src/UniswapV2/UniswapV2ERC20.sol:92]: what's the point of returning `true`?

File: solidity_contracts/src/UniswapV2/test/ERC20.sol [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
+import "../UniswapV2ERC20.sol";
 
 contract ERC20 is UniswapV2ERC20 {
     constructor(uint _totalSupply) public {
         _mint(msg.sender, _totalSupply);
     }
 }
+
+// TODO: Fix address collision if token_a and token_b fixtures both use UniswapV2/ERC20.sol::ERC20
+contract ERC20Bis is UniswapV2ERC20 {
```
Comment (ClementWalter) [inline solidity_contracts/src/UniswapV2/test/ERC20.sol:12]: I see no reason to have this

File: solidity_contracts/src/UniswapV2/UniswapV2ERC20.sol [hunk 6]
```diff
@@ -87,6 +87,11 @@ contract UniswapV2ERC20 is IUniswapV2ERC20 {
         return true;
     }
 
+    function mint(address to, uint value) external returns (bool) {
```
Comment (ClementWalter) [inline solidity_contracts/src/UniswapV2/UniswapV2ERC20.sol:90]: I think that this file is exactly the one in the univ2 library, better update the ERC20.sol imho

File: solidity_contracts/src/UniswapV2Router/WETH9.sol
Comment (ClementWalter) [inline solidity_contracts/src/UniswapV2Router/WETH9.sol:1]: I don't think that WETH9 belongs to Univ2 folder

File: tests/end_to_end/UniswapV2/conftest.py [hunk 7]
```diff
@@ -17,12 +17,21 @@ async def token_a(deploy_solidity_contract, owner):
 async def token_b(deploy_solidity_contract, owner):
     return await deploy_solidity_contract(
         "UniswapV2",
-        "ERC20",
+        "ERC20Bis",
```
Comment (ClementWalter) [inline tests/end_to_end/UniswapV2/conftest.py:20]: I don't understand why this is needed

File: tests/end_to_end/UniswapV2/test_uniswap_v2_router.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+import pytest
+
+
+@pytest.mark.asyncio(scope="session")
+# @pytest.mark.UniswapV2Router
```
Comment (ClementWalter) [inline tests/end_to_end/UniswapV2/test_uniswap_v2_router.py:5]: why not putting a mark?

File: tests/end_to_end/UniswapV2/test_uniswap_v2_router.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+class TestUniswapV2Router:
+    class TestDeploy:
+        async def test_should_set_constants(self, router, weth, factory):
+            assert await router.WETH() == weth.address
+            print("WETH Address", weth.address)
```
Comment (ClementWalter) [inline tests/end_to_end/UniswapV2/test_uniswap_v2_router.py:10]: if you print, I think that you don't know how to use a debugger
1. select python interpreter
2. configure test
3. debug test

File: tests/end_to_end/UniswapV2/test_uniswap_v2_router.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            assert await router.factory() == factory.address
+
+    class TestAddLiquidity:
+        async def test_should_add_liquidity(self, router, token_b, token_a, owner):
+            amount_A_desired = 1000 * 10**18  # This needs to match the token's decimals
```
Comment (ClementWalter) [inline tests/end_to_end/UniswapV2/test_uniswap_v2_router.py:15]: you can query `token_a.decimals()` instead of hard coding the value

**Review by ClementWalter — CHANGES_REQUESTED** [review]

File: Makefile (hunk 3 above)
Comment (Eikix) [reply Makefile]: Modifying this to be the requested prod setting: 
6M val. / 14M exec

File: kakarot_scripts/utils/kakarot.py [hunk 11]
```diff
@@ -380,7 +379,7 @@ async def eth_send_transaction(
     transaction_events = [
         event
         for event in receipt.events
-        if event.from_address == evm_account.address
+        if event.from_address in [0, evm_account.address]
```
Comment (enitrat) [inline kakarot_scripts/utils/kakarot.py:382]: restore

File: solidity_contracts/src/UniswapV2/test/ERC20.sol [hunk 12]
```diff
@@ -1,9 +1,13 @@
...
 
-import '../UniswapV2ERC20.sol';
+import "../UniswapV2ERC20.sol";
 
 contract ERC20 is UniswapV2ERC20 {
     constructor(uint _totalSupply) public {
         _mint(msg.sender, _totalSupply);
     }
+
+    function mint(address to, uint value) external returns (bool) {
```
Comment (enitrat) [inline solidity_contracts/src/UniswapV2/test/ERC20.sol:10]: so it returns always false ?

Comment (enitrat) [reply solidity_contracts/src/UniswapV2/test/ERC20.sol:10]: I think in OZ's spec it just always returns true

Comment (Eikix) [reply solidity_contracts/src/UniswapV2/test/ERC20.sol:10]: ah oop

**Review by enitrat — DISMISSED** [review]
lgtm

**Comment by enitrat** [comment]
do we need all these tests? I thought the purpose was to evaluate the steps taken by addLiquidity ?

**Review by enitrat — DISMISSED** [review]

**Review by enitrat — APPROVED** [review]

**Pull request merged**